### PR TITLE
[FLINK-21985][table-api] Support Explain Query/Modifcation syntax in Calcite Parser

### PR DESCRIPTION
**Pull request by chaozwn**
…Calcite Parser

## What is the purpose of the change

*(For example: This pull request makes task deployment go through the blob server, rather than through RPC. That way we avoid re-transferring them on each deployment (during recovery).)*

## Brief change log

*(for example:)*
  - *The TaskInfo is stored in the blob store on job creation time as a persistent artifact*
  - *Deployments RPC transmits only the blob storage reference*
  - *TaskManagers retrieve the TaskInfo from the blob cache*

## Verifying this change

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (100MB)*
  - *Extended integration test for recovery after master (JobManager) failure*
  - *Added test that validates that TaskInfo is transferred only once across recoveries*
  - *Manually verified the change by running a 4 node cluser with 2 JobManagers and 4 TaskManagers, a stateful streaming program, and killing one JobManager and two TaskManagers during the execution, verifying that recovery happens correctly.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (yes / no / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-table/flink-sql-parser-hive/src/main/codegen/includes/parserImpls.ftl [hunk 1]
```diff
@@ -1605,3 +1605,17 @@ SqlShowModules SqlShowModules() :
...
+SqlNode SqlRichExplain() :
+{
+    SqlNode stmt;
+}
+{
+    <EXPLAIN> (<PLAN> <FOR>)*
+    stmt = SqlQueryOrDml() {
+        return new SqlRichExplain(getPos(),stmt);
+    }
+}
```
Comment (fsk119) [inline flink-table/flink-sql-parser-hive/src/main/codegen/includes/parserImpls.ftl:1621]: Add a new line

File: flink-table/flink-sql-parser-hive/src/test/java/org/apache/flink/sql/parser/hive/FlinkHiveSqlParserImplTest.java [hunk 2]
```diff
@@ -482,4 +482,54 @@ public void testShowModules() {
...
+
+    @Test
+    public void testExplainAsXml() {
+        // unsupport testExplainWithType now
+    }
+
+    @Test
+    public void testExplainAsJson() {
+        // unsupport testExplainWithType now
+    }
```
Comment (fsk119) [inline flink-table/flink-sql-parser-hive/src/test/java/org/apache/flink/sql/parser/hive/FlinkHiveSqlParserImplTest.java:521]: Add comment: 
// TODO: FLINK-20562

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlRichExplain.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        writer.keyword("EXPLAIN");
+        statement.unparse(writer, leftPrec, rightPrec);
+    }
+
+    @Override
+    public void setOperand(int i, SqlNode operand) {
+        if (i == 0) {
+            statement = operand;
+        } else {
+            throw new AssertionError(i);
```
Comment (fsk119) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlRichExplain.java:70]: Please use meaningful exception message.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/calcite/FlinkPlannerImpl.scala [hunk 4]
```diff
@@ -146,6 +146,10 @@ class FlinkPlannerImpl(
           val validated = validator.validate(explain.getExplicandum)
           explain.setOperand(0, validated)
           explain
+        case richExplain: SqlRichExplain =>
+          val validated = validator.validate(richExplain.getStatement)
+          richExplain.setOperand(0,validated)
+          richExplain
```
Comment (fsk119) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/calcite/FlinkPlannerImpl.scala:152]: Move forward before `SqlExplain`?  Do we really need SqlExplain?

File: flink-table/flink-sql-parser-hive/src/test/java/org/apache/flink/sql/parser/hive/FlinkHiveSqlParserImplTest.java [hunk 5]
```diff
@@ -482,4 +482,54 @@ public void testShowModules() {
...
+
+    @Test
+    public void testExplainWithType() {
+        // unsupport testExplainWithType now
+    }
+
+    @Test
+    public void testExplainAsXml() {
+        // unsupport testExplainWithType now
+    }
```
Comment (fsk119) [inline flink-table/flink-sql-parser-hive/src/test/java/org/apache/flink/sql/parser/hive/FlinkHiveSqlParserImplTest.java:516]: Add comment:
Don't supported feature. Escape the test.

File: flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java [hunk 6]
```diff
@@ -1215,6 +1214,56 @@ public void testEnd() {
...
+    }
+
+    @Test
+    public void testExplainAsXml() {
+        // unsupport testExplainWithType now
+    }
+
+    @Test
+    public void testExplainAsJson() {
+        // unsupport testExplainWithType now
```
Comment (fsk119) [inline flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java:1251]: ditto

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ExplainOperation.java [hunk 7]
```diff
@@ -21,8 +21,8 @@
 import java.util.Collections;
 
 /**
- * Operation to describe an EXPLAIN statement. NOTES: currently, only default behavior(EXPLAIN PLAN
- * FOR xx) is supported.
+ * Operation to describe an EXPLAIN statement. NOTES: currently, only default behavior(EXPLAIN xx)
+ * is supported.
```
Comment (fsk119) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ExplainOperation.java:25]: Explain plan for is also supported.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/calcite/FlinkPlannerImpl.scala [hunk 8]
```diff
@@ -18,28 +18,26 @@
...
+import org.apache.calcite.sql.advise.SqlAdvisorValidator
 import org.apache.calcite.sql.validate.SqlValidator
 import org.apache.calcite.sql.{SqlExplain, SqlKind, SqlNode, SqlOperatorTable}
 import org.apache.calcite.sql2rel.{SqlRexConvertletTable, SqlToRelConverter}
 import org.apache.calcite.tools.{FrameworkConfig, RelConversionException}
+import org.apache.flink.sql.parser.ExtendedSqlNode
+import org.apache.flink.sql.parser.dql._
+import org.apache.flink.table.api.{TableException, ValidationException}
+import org.apache.flink.table.catalog.CatalogReader
+import org.apache.flink.table.parse.CalciteParser
```
Comment (fsk119) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/calcite/FlinkPlannerImpl.scala:39]: Rollback

**Review by fsk119 — COMMENTED** [review]
Thanks for your conribution. I left some comments.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/calcite/FlinkPlannerImpl.scala [hunk 9]
```diff
@@ -94,9 +96,9 @@ class FlinkPlannerImpl(
       catalogReader,
       typeFactory,
       SqlValidator.Config.DEFAULT
-        .withIdentifierExpansion(true)
-        // Disable implicit type coercion for now.
-        .withTypeCoercionEnabled(false))
+          .withIdentifierExpansion(true)
+          // Disable implicit type coercion for now.
+          .withTypeCoercionEnabled(false))
```
Comment (fsk119) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/calcite/FlinkPlannerImpl.scala]: Please use 2 space in scala file.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/calcite/FlinkPlannerImpl.scala [hunk 10]
```diff
@@ -181,7 +183,7 @@ class FlinkPlannerImpl(
       queryString: String,
       schemaPath: util.List[String],
       viewPath: util.List[String])
-  : RelRoot = {
+    : RelRoot = {
```
Comment (fsk119) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/calcite/FlinkPlannerImpl.scala]: roll back

File: flink-table/flink-sql-parser-hive/src/test/java/org/apache/flink/sql/parser/hive/FlinkHiveSqlParserImplTest.java [hunk 11]
```diff
@@ -482,4 +482,54 @@ public void testShowModules() {
...
+    @Test
+    public void testExplain() {
+        String sql = "explain plan for select * from emps";
+        String expected = "EXPLAIN SELECT *\n" + "FROM `EMPS`";
+        this.sql(sql).ok(expected);
+    }
+
+    @Test
+    public void testExplainJsonFormat() {
+        // Don't supported feature. Escape the test.
```
Comment (fsk119) [inline flink-table/flink-sql-parser-hive/src/test/java/org/apache/flink/sql/parser/hive/FlinkHiveSqlParserImplTest.java:495]: Unsupported feature.

File: flink-table/flink-sql-client/src/test/resources/sql/table.q [hunk 12]
```diff
@@ -204,3 +204,119 @@ drop table `mod`;
...
+# ==========================================================================
+
+CREATE TABLE IF NOT EXISTS orders (
+ `user` BIGINT NOT NULl,
+ product VARCHAR(32),
+ amount INT,
+ ts TIMESTAMP(3),
+ ptime AS PROCTIME(),
+ PRIMARY KEY(`user`) NOT ENFORCED,
+ WATERMARK FOR ts AS ts - INTERVAL '1' SECONDS
```
Comment (fsk119) [inline flink-table/flink-sql-client/src/test/resources/sql/table.q:219]: please style the DDL as above. Add "  " before fields

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ExplainOperation.java [hunk 13]
```diff
@@ -21,8 +21,8 @@
 import java.util.Collections;
 
 /**
- * Operation to describe an EXPLAIN statement. NOTES: currently, only default behavior(EXPLAIN PLAN
- * FOR xx) is supported.
+ * Operation to describe an EXPLAIN statement. NOTES: currently, only default behavior(EXPLAIN [plan
+ * for]* xx) is supported.
```
Comment (fsk119) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ExplainOperation.java:25]: nit: Use UPPSER CASE "PLAN FOR"

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/stream/StreamTableEnvironmentTest.scala [hunk 14]
```diff
@@ -322,6 +322,10 @@ class StreamTableEnvironmentTest extends TableTestBase {
     } catch {
       case e: TableException =>
         assertTrue(e.getMessage.contains("Only default behavior is supported now"))
```
Comment (fsk119) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/stream/StreamTableEnvironmentTest.scala:324]: Remove this. It seems no one will getinto this branch? In my local environment, the test passes.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/batch/BatchTableEnvironmentTest.scala [hunk 15]
```diff
@@ -542,6 +542,9 @@ class BatchTableEnvironmentTest extends TableTestBase {
     } catch {
       case e: TableException =>
         assertTrue(e.getMessage.contains("Only default behavior is supported now"))
```
Comment (fsk119) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/batch/BatchTableEnvironmentTest.scala:544]: Remove this. It seems no one will getinto this branch? In my local environment, the test passes.

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlRichExplain.java [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlOperator;
+import org.apache.calcite.sql.SqlSpecialOperator;
+import org.apache.calcite.sql.SqlWriter;
+import org.apache.calcite.sql.parser.SqlParserPos;
+
+import java.util.Collections;
+import java.util.List;
+
+/** EXPLAIN (plan for)* STATEMENT sql call. */
```
Comment (fsk119) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlRichExplain.java:32]: Use upper case.

**Review by fsk119 — COMMENTED** [review]
Look good to me in general.

**Review by fsk119 — APPROVED** [review]
LGTM. CC @wuchong

**Review by wuchong — APPROVED** [review]
LGTM.

**Pull request closed without merging**